8301584: Generational ZGC: Add barrier elision tests

robcasloz opened the pull request:
This changeset adds test cases exercising C2's barrier elision on combinations of different

- pairs of dominating / dominated accesses (allocations, loads, stores, atomic operations),
- control structures (local, conditions, loops),
- object types (arrays, class instances), and
- levels of information available during C2's analysis (known vs. unknown array indices).

The changeset exposes node barrier data to the [IR test framework](https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/README.md) by including it in the node's dump output. This extension is useful not just for testing but also for debugging purposes, for example when exploring C2's intermediate representation with IGV:

![igv](https://user-images.githubusercontent.com/8792647/216639224-aa288dd8-72cc-43a5-a445-a93d65004dac.png)

The following tests illustrate a missing optimization in C2's barrier elision (reported in [JDK-8301769](https://bugs.openjdk.org/browse/JDK-8301769)), where barriers that should be elided are not:

- `testStoreThenAtomic`
- `testAtomicThenLoad`
- `testAtomicThenStore`
- `testAtomicThenAtomic`

The changeset does not contain IR checks expecting elision in these cases, to reduce CI pipeline noise. The tests are restricted by now to x64, since volatile loads and stores in other platforms suffer from the same issue (see [JDK-8301769](https://bugs.openjdk.org/browse/JDK-8301769)).

**Testing:** tier1-7 (windows-x64, linux-x64, macosx-x64; release and debug mode)

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301584](https://bugs.openjdk.org/browse/JDK-8301584): Generational ZGC: Add barrier elision tests


### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [3a5ef481](https://git.openjdk.org/zgc/pull/12/files/3a5ef48136d90548f59ac7d33fa235e0777a0de2)
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - Committer) ⚠️ Review applies to [3a5ef481](https://git.openjdk.org/zgc/pull/12/files/3a5ef48136d90548f59ac7d33fa235e0777a0de2)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/zgc pull/12/head:pull/12` \
`$ git checkout pull/12`

Update a local copy of the PR: \
`$ git checkout pull/12` \
`$ git pull https://git.openjdk.org/zgc pull/12/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12`

View PR using the GUI difftool: \
`$ git pr show -t 12`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/zgc/pull/12.diff">https://git.openjdk.org/zgc/pull/12.diff</a>

</details>

robcasloz committed:
Dump MachNode barrier data when available

robcasloz committed:
Add basic barrier elision tests

robcasloz committed:
Fix merge typo

robcasloz committed:
Update barrier elision tests after latest blackhole refinements

robcasloz committed:
Remove unnecessary volatile qualifiers

robcasloz committed:
Simplify

robcasloz committed:
Add allocation tests

robcasloz committed:
Add array access tests

robcasloz committed:
Add different field tests

robcasloz committed:
Add control-flow tests

robcasloz committed:
Add atomic operation tests

robcasloz committed:
Replace uses of Unsafe with var handles

robcasloz committed:
Complete atomic operation tests

robcasloz committed:
Simplify tests

robcasloz committed:
Remove unused arguments

robcasloz committed:
Update copyright years

robcasloz committed:
Remove failing IR checks to avoid CI pipeline noise

robcasloz committed:
Remove unnecessary keywords

robcasloz committed:
Match also aarch64 memory access node names

robcasloz committed:
Restrict tests to x64 by now

fisk reviewed (approved):
Nice test! Looks good.

robcasloz commented:
Thanks for reviewing, Erik!

xmas92 commented on test/hotspot/jtreg/compiler/gcbarriers/TestZGCBarrierElision.java:
```diff
@@ -0,0 +1,359 @@
+    }
+
+    @Test
+    // The second atomic access barrier should be elided, but is not.
```
Is it worth explicitly documenting the behaviour and the wanted/expected behaviour in annotations?
E.g.
```java
    @Test
    @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, REMAINING, "2" }, phase = CompilePhase.FINAL_CODE)
    @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, ELIDED, "0" }, phase = CompilePhase.FINAL_CODE)
    // The second atomic access barrier should be elided, but is not.
    // @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, REMAINING, "1" }, phase = CompilePhase.FINAL_CODE)
    // @IR(counts = { IRNode.Z_GET_AND_SET_P_WITH_BARRIER_FLAG, ELIDED, "1" }, phase = CompilePhase.FINAL_CODE)
    static void testAtomicThenAtomic(Outer o, Inner i) {
        field1VarHandle.getAndSet​(o, i);
        field1VarHandle.getAndSet​(o, i);
    }
```

Same goes for all the other test cases.

robcasloz replied:
IR checks in compiler tests are commonly interpreted as documentation for expected behavior only. I fear adding IR checks for the current (unexpected) behavior would risk misleading inattentive readers, so I would rather leave it as-is if you don't mind.

xmas92 reviewed (approved):
Looks good. Interesting test framework, good to know this exists.

We should upstream the `src/hotspot/share/gc/shared/c2/barrierSetC2.hpp` and `src/hotspot/share/opto/machnode.cpp` changes

xmas92 commented on src/hotspot/share/opto/machnode.cpp:
```diff
@@ -23,6 +23,7 @@
  */
 
 #include "precompiled.hpp"
+#include "gc/shared/barrierSet.hpp"
```
Seems to need `#include "gc/shared/c2/barrierSetC2.hpp"` on some platforms (just looking at your GHA)

robcasloz replied:
Good catch, done!

xmas92 commented on src/hotspot/share/gc/shared/c2/barrierSetC2.hpp:
```diff
@@ -30,6 +30,7 @@
 #include "opto/loopnode.hpp"
 #include "opto/matcher.hpp"
 #include "opto/memnode.hpp"
+#include "opto/machnode.hpp"
```
Include order

robcasloz replied:
Done.

robcasloz committed:
Fix include order

robcasloz committed:
Add missing include

robcasloz commented:
Thanks for reviewing, Axel! I just addressed your feedback, [GHA builds](https://github.com/robcasloz/zgc/actions/runs/4122408512/jobs/7120298560) look good now. Please let me know if there is anything else that should be done before integrating.

fisk commented:
Ship it!

robcasloz commented:
/integrate
